src/Microsoft.Extensions.Configuration.AzureAppConfiguration/ConfigurationClientManager.cs

```diff
@@ -118,8 +255,146 @@ public Uri GetEndpointForClient(ConfigurationClient client)
-            
+
+            if (_dynamicClients != null && currentClient == null)
+            {
+                currentClient = _dynamicClients.FirstOrDefault(c => c.Client == client);
+            }
+
             return currentClient?.Endpoint;
         }
+
+        public void SetLogger(Logger logger)
+        {
+            if (logger == null)
+            {
+                throw new ArgumentNullException(nameof(logger));
+            }
+
+            _logger = logger;
+        }
+
+        private async Task OptimisticRefreshFallbackClients(DateTimeOffset now, CancellationToken cancellationToken)
+        {
+            if (_replicaDiscoveryEnabled &&
+                now >= _lastFallbackClientRefreshAttempt + MinimalClientRefreshInterval &&
+                (_dynamicClients == null ||
+                now >= _lastFallbackClientRefresh + FallbackClientRefreshExpireInterval))
+            {
+                _lastFallbackClientRefreshAttempt = now;
+
+                using var cts = CancellationTokenSource.CreateLinkedTokenSource(cancellationToken);
+
+                cts.CancelAfter(MinimalClientRefreshInterval);
+
+                try
+                {
+                    await RefreshFallbackClients(cts.Token).ConfigureAwait(false);
+                }
+                catch (OperationCanceledException) when (!cancellationToken.IsCancellationRequested)
+                {
+                    // Do nothing if origin cancellationToken is not cancelled.
+                }
+            }
+        }
+
+        private async Task RefreshFallbackClients(CancellationToken cancellationToken)
+        {
+            IEnumerable<string> srvTargetHosts = Enumerable.Empty<string>();
+
+            try
+            {
+                srvTargetHosts = await _srvLookupClient.QueryAsync(_endpoint.DnsSafeHost, cancellationToken).ConfigureAwait(false);
+            }
+            // Catch and log all exceptions thrown by srv lookup client to avoid new possible exceptions on app startup.
+            catch (SocketException ex)
+            {
+                _logger.LogWarning(LogHelper.BuildFallbackClientLookupFailMessage(ex.Message));
+                return;
+            }
+            catch (DnsResponseException ex)
+            {
+                _logger.LogWarning(LogHelper.BuildFallbackClientLookupFailMessage(ex.Message));
+                return;
+            }
+            catch (InvalidOperationException ex)
+            {
+                _logger.LogWarning(LogHelper.BuildFallbackClientLookupFailMessage(ex.Message));
+                return;
+            }
+            catch (DnsXidMismatchException ex)
+            {
+                _logger.LogWarning(LogHelper.BuildFallbackClientLookupFailMessage(ex.Message));
+                return;
+            }
+
+            _dynamicClients ??= new List<ConfigurationClientWrapper>();
+
+            // Shuffle the results to ensure hosts can be picked randomly.
```

Comment:
There is RFC protocol how to resolve service instances for DNS-Based Service Discovery. It's not just random.

Reply:
Updated, random-based, but honor with the DNS-Based Service Discovery protocol